core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java

```diff
@@ -0,0 +1,147 @@
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * If a master is detected as a standby master. It will set up its MetaMasterSync and manage
+ * its own {@link MetaMasterMasterClient} which helps communicate with the leader master.
+ *
+ * When running, the standby master will send its heartbeat to the leader master.
+ * The leader master may respond to the heartbeat with a command which will be executed.
+ * After which, the task will wait for the elapsed time since its last heartbeat
+ * has reached the heartbeat interval. Then the cycle will continue.
+ */
+@NotThreadSafe
+public final class MetaMasterSync implements HeartbeatExecutor {
+  private static final Logger LOG = LoggerFactory.getLogger(MetaMasterSync.class);
+
+  /**
+   * The ID of this standby master.
+   * This may change if the leader master asks the standby master to re-register.
+   */
+  private final AtomicReference<Long> mMasterId;
+
+  /** The hostname of this standby master. */
+  private final String mMasterHostname;
+
+  /** Milliseconds between heartbeats before a timeout. */
+  private final int mHeartbeatTimeoutMs;
+
+  /** Client for communication with the leader master. */
+  private final MetaMasterMasterClient mMasterClient;
+
+  /** Last System.currentTimeMillis() timestamp when a heartbeat successfully completed. */
+  private long mLastSuccessfulHeartbeatMs;
+
+  /**
+   * Creates a new instance of {@link MetaMasterSync}.
+   *
+   * @param masterId the master id
+   * @param masterHostname the master hostname
+   * @param masterClient the meta master client
+   */
+  public MetaMasterSync(AtomicReference<Long>  masterId,
+      String masterHostname, MetaMasterMasterClient masterClient) throws IOException {
+    mMasterId = masterId;
+    mMasterHostname = masterHostname;
+    mMasterClient = masterClient;
+    mHeartbeatTimeoutMs = (int) Configuration.getMs(PropertyKey.MASTER_HEARTBEAT_TIMEOUT_MS);
+
+    registerWithMaster();
+    mLastSuccessfulHeartbeatMs = System.currentTimeMillis();
+  }
+
+  /**
+   * Registers with the Alluxio master. This should be called
+   * before the continuous heartbeat thread begins.
+   */
+  private void registerWithMaster() throws IOException {
+    mMasterClient.register(mMasterId.get(),
+        Configuration.getConfiguration(PropertyKey.Scope.MASTER));
+  }
+
+  /**
+   * Heartbeats to the leader master node.
+   */
+  @Override
+  public void heartbeat() {
+    MetaCommand command = null;
+    try {
+      command = mMasterClient.heartbeat(mMasterId.get());
+      handleCommand(command);
+      mLastSuccessfulHeartbeatMs = System.currentTimeMillis();
+    } catch (IOException e) {
+      // An error occurred, log and ignore it or error if heartbeat timeout is reached
+      if (command == null) {
+        LOG.error("Failed to receive leader master heartbeat command.", e);
+      } else {
+        LOG.error("Failed to receive or execute leader master heartbeat command: {}",
+            command.toString(), e);
```

Comment:
nit: unnecessary `command.toString()` but `command` should be fine

Reply:
Done, thanks!